# IEEE-CIS Fraud Detection

## Notebook 1: Data Cleaning

This notebook loads the raw transaction and identity data, examines its structure,
handles missing values, fixes data types, and produces a clean dataset ready for
exploratory analysis.

**Dataset:** IEEE-CIS Fraud Detection (Kaggle)  
**Rows:** ~590,000 transactions  
**Target:** isFraud (1 = fraud, 0 = legitimate)

In [1]:
import pandas as pd
import numpy as np

# suppress pandas warnings that clutter the output
pd.set_option('display.max_columns', 100)
pd.set_option('display.float_format', lambda x: '%.3f' % x)

## 1. Load the Data

We load the transaction and identity tables separately and check their shapes
before doing anything else. This tells us how many rows and columns we are
working with.

In [2]:
transaction = pd.read_csv('/kaggle/input/competitions/ieee-fraud-detection/train_transaction.csv')
identity = pd.read_csv('/kaggle/input/competitions/ieee-fraud-detection/train_identity.csv')

df = transaction.merge(identity, on='TransactionID', how='left')

print(f'Transaction table : {transaction.shape}')
print(f'Identity table    : {identity.shape}')
print(f'Merged table      : {df.shape}')

Transaction table : (590540, 394)
Identity table    : (144233, 41)
Merged table      : (590540, 434)


In [3]:
df.head()

,TransactionID,isFraud,TransactionDT,TransactionAmt,ProductCD,card1,card2,card3,card4,card5,card6,addr1,addr2,dist1,dist2,P_emaildomain,R_emaildomain,C1,C2,C3,C4,C5,C6,C7,C8,C9,C10,C11,C12,C13,C14,D1,D2,D3,D4,D5,D6,D7,D8,D9,D10,D11,D12,D13,D14,D15,M1,M2,M3,M4,...,V330,V331,V332,V333,V334,V335,V336,V337,V338,V339,id_01,id_02,id_03,id_04,id_05,id_06,id_07,id_08,id_09,id_10,id_11,id_12,id_13,id_14,id_15,id_16,id_17,id_18,id_19,id_20,id_21,id_22,id_23,id_24,id_25,id_26,id_27,id_28,id_29,id_30,id_31,id_32,id_33,id_34,id_35,id_36,id_37,id_38,DeviceType,DeviceInfo
0,2987000,0,86400,68.500,W,13926,NaN,150.000,discover,142.000,credit,315.000,87.000,19.000,NaN,NaN,NaN,1.000,1.000,0.000,0.000,0.000,1.000,0.000,0.000,1.000,0.000,2.000,0.000,1.000,1.000,14.000,NaN,13.000,NaN,NaN,NaN,NaN,NaN,NaN,13.000,13.000,NaN,NaN,NaN,0.000,T,T,T,M2,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
1,2987001,0,86401,29.000,W,2755,404.000,150.000,mastercard,102.000,credit,325.000,87.000,NaN,NaN,gmail.com,NaN,1.000,1.000,0.000,0.000,0.000,1.000,0.000,0.000,0.000,0.000,1.000,0.000,1.000,1.000,0.000,NaN,NaN,0.000,NaN,NaN,NaN,NaN,NaN,0.000,NaN,NaN,NaN,NaN,0.000,NaN,NaN,NaN,M0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
2,2987002,0,86469,59.000,W,4663,490.000,150.000,visa,166.000,debit,330.000,87.000,287.000,NaN,outlook.com,NaN,1.000,1.000,0.000,0.000,0.000,1.000,0.000,0.000,1.000,0.000,1.000,0.000,1.000,1.000,0.000,NaN,NaN,0.000,NaN,NaN,NaN,NaN,NaN,0.000,315.000,NaN,NaN,NaN,315.000,T,T,T,M0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
3,2987003,0,86499,50.000,W,18132,567.000,150.000,mastercard,117.000,debit,476.000,87.000,NaN,NaN,yahoo.com,NaN,2.000,5.000,0.000,0.000,0.000,4.000,0.000,0.000,1.000,0.000,1.000,0.000,25.000,1.000,112.000,112.000,0.000,94.000,0.000,NaN,NaN,NaN,NaN,84.000,NaN,NaN,NaN,NaN,111.000,NaN,NaN,NaN,M0,...,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN
4,2987004,0,86506,50.000,H,4497,514.000,150.000,mastercard,102.000,credit,420.000,87.000,NaN,NaN,gmail.com,NaN,1.000,1.000,0.000,0.000,0.000,1.000,0.000,1.000,0.000,1.000,1.000,0.000,1.000,1.000,0.000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,...,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,0.000,70787.000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,100.000,NotFound,NaN,-480.000,New,NotFound,166.000,NaN,542.000,144.000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,New,NotFound,Android 7.0,samsung browser 6.2,32.000,2220x1080,match_status:2,T,F,T,T,mobile,SAMSUNG SM-G892A Build/NRD90M


## 2. Memory Optimization

The merged dataframe occupies 2.5 GB using default int64 and float64 types.
Casting to int32 and float32 halves memory usage with no loss of precision
for this dataset.

In [4]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 590540 entries, 0 to 590539
Columns: 434 entries, TransactionID to DeviceInfo
dtypes: float64(399), int64(4), object(31)
memory usage: 2.5 GB


In [5]:
# float64 uses 8 bytes per value — float32 uses 4 bytes with sufficient precision for this data
for col in df.select_dtypes(include='float64').columns:
    df[col] = df[col].astype('float32')

# check how much memory we saved
for col in df.select_dtypes(include='int64').columns:
    df[col] = df[col].astype('int32')

print(f'Memory usage after optimization: {df.memory_usage(deep=True).sum() / 1e9:.2f} GB')

Memory usage after optimization: 1.68 GB


## 3. Missing Values

414 out of 434 columns have missing values. Identity columns are missing for
transactions where no identity data was collected — this is expected.

Columns missing more than 90% of values are dropped. For the rest, numerical
columns are filled with the median and categorical columns with 'unknown'.

In [6]:
missing = df.isnull().sum()
missing_pct = (missing / len(df)) * 100
missing_df = pd.DataFrame({
    'missing_count': missing,
    'missing_pct': missing_pct
}).query('missing_count > 0').sort_values('missing_pct', ascending=False)

print(f'Columns with missing values: {len(missing_df)} out of {df.shape[1]}')
print(missing_df.head(80))

Columns with missing values: 414 out of 434
       missing_count  missing_pct
id_24         585793       99.196
id_25         585408       99.131
id_07         585385       99.127
id_08         585385       99.127
id_21         585381       99.126
...              ...          ...
id_13         463220       78.440
id_16         461200       78.098
V242          460110       77.913
V241          460110       77.913
V278          460110       77.913

[80 rows x 2 columns]


In [7]:
# columns missing more than 90% of values carry no meaningful signal
cols_to_drop = missing_df[missing_df['missing_pct'] > 90].index.tolist()

df.drop(columns=cols_to_drop, inplace=True)

print(f'Columns dropped : {len(cols_to_drop)}')
print(f'Remaining columns: {df.shape[1]}')

Columns dropped : 12
Remaining columns: 422


In [8]:
df.info(memory_usage='deep')

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 590540 entries, 0 to 590539
Columns: 422 entries, TransactionID to DeviceInfo
dtypes: float32(389), int32(4), object(29)
memory usage: 1.5 GB


In [9]:
fraud_counts = df['isFraud'].value_counts()
fraud_pct = df['isFraud'].value_counts(normalize=True) * 100

print(fraud_counts)
print()
print(f'Legitimate transactions: {fraud_pct[0]:.2f}%')
print(f'Fraudulent transactions: {fraud_pct[1]:.2f}%')

isFraud
0    569877
1     20663
Name: count, dtype: int64

Legitimate transactions: 96.50%
Fraudulent transactions: 3.50%


In [10]:
# numerical columns: fill with median because median is robust to outliers
# categorical columns: fill with the string 'unknown' to preserve the fact that data was missing

num_cols = df.select_dtypes(include=['float32', 'int32']).columns
cat_cols = df.select_dtypes(include='object').columns
df[num_cols] = df[num_cols].fillna(df[num_cols].median())
df[cat_cols] = df[cat_cols].fillna('unknown')

print(f'Remaining missing values: {df.isnull().sum().sum()}')


Remaining missing values: 0


## 4. Data Type Validation

We verify that no numeric column contains hidden string values and no categorical
column contains numeric values stored as strings.

In [11]:
print(df.dtypes.value_counts())
print()
print('Object columns:')
print(df.select_dtypes(include='object').columns.tolist())

float32    389
object      29
int32        4
Name: count, dtype: int64

Object columns:
['ProductCD', 'card4', 'card6', 'P_emaildomain', 'R_emaildomain', 'M1', 'M2', 'M3', 'M4', 'M5', 'M6', 'M7', 'M8', 'M9', 'id_12', 'id_15', 'id_16', 'id_28', 'id_29', 'id_30', 'id_31', 'id_33', 'id_34', 'id_35', 'id_36', 'id_37', 'id_38', 'DeviceType', 'DeviceInfo']


In [12]:
for col in df.select_dtypes(include='object').columns:
    print(f'{col}: {df[col].unique()[:5]}')

ProductCD: ['W' 'H' 'C' 'S' 'R']
card4: ['discover' 'mastercard' 'visa' 'american express' 'unknown']
card6: ['credit' 'debit' 'unknown' 'debit or credit' 'charge card']
P_emaildomain: ['unknown' 'gmail.com' 'outlook.com' 'yahoo.com' 'mail.com']
R_emaildomain: ['unknown' 'gmail.com' 'hotmail.com' 'outlook.com' 'anonymous.com']
M1: ['T' 'unknown' 'F']
M2: ['T' 'unknown' 'F']
M3: ['T' 'unknown' 'F']
M4: ['M2' 'M0' 'unknown' 'M1']
M5: ['F' 'T' 'unknown']
M6: ['T' 'F' 'unknown']
M7: ['unknown' 'F' 'T']
M8: ['unknown' 'F' 'T']
M9: ['unknown' 'F' 'T']
id_12: ['unknown' 'NotFound' 'Found']
id_15: ['unknown' 'New' 'Found' 'Unknown']
id_16: ['unknown' 'NotFound' 'Found']
id_28: ['unknown' 'New' 'Found']
id_29: ['unknown' 'NotFound' 'Found']
id_30: ['unknown' 'Android 7.0' 'iOS 11.1.2' 'Mac OS X 10_11_6' 'Windows 10']
id_31: ['unknown' 'samsung browser 6.2' 'mobile safari 11.0' 'chrome 62.0'
 'chrome 62.0 for android']
id_33: ['unknown' '2220x1080' '1334x750' '1280x800' '1366x768']
id_34: ['unkn

In [13]:
# check if any object column contains numeric values hiding as strings
for col in df.select_dtypes(include='object').columns:
    try:
        converted = pd.to_numeric(df[col], errors='coerce')
        non_null_pct = converted.notna().sum() / len(df) * 100
        if non_null_pct > 50:
            print(f'{col}: {non_null_pct:.1f}% of values are numeric — consider converting')
    except:
        pass

print('Check complete')

Check complete


In [14]:
# read a small sample of the raw files as plain strings to catch any numeric column hiding string values
raw_sample = pd.read_csv(
    '/kaggle/input/competitions/ieee-fraud-detection/train_transaction.csv',
    nrows=5,
    dtype=str
)

for col in raw_sample.columns:
    vals = raw_sample[col].unique()
    # if a column has non-numeric non-null values it was never truly numeric
    for v in vals:
        try:
            float(v)
        except (ValueError, TypeError):
            if v != 'nan':
                print(f'{col}: contains string value -> "{v}"')
                break

ProductCD: contains string value -> "W"
card4: contains string value -> "discover"
card6: contains string value -> "credit"
P_emaildomain: contains string value -> "gmail.com"
M1: contains string value -> "T"
M2: contains string value -> "T"
M3: contains string value -> "T"
M4: contains string value -> "M2"
M5: contains string value -> "F"
M6: contains string value -> "T"
M7: contains string value -> "F"
M8: contains string value -> "F"
M9: contains string value -> "F"


In [15]:
df.to_csv('cleaned_data.csv', index=False)

print(f'Saved: cleaned_data.csv')
print(f'Shape: {df.shape}')

Saved: cleaned_data.csv
Shape: (590540, 422)


## 5. Summary

| Step | Result |
|---|---|
| Original shape | (590540, 434) |
| Columns dropped (>90% missing) | 12 |
| Final shape | (590540, 422) |
| Missing values remaining | 0 |
| Memory usage | 1.68 GB |
| Fraud rate | 3.50% |

The cleaned dataset is saved as `cleaned_data.csv` and is ready for exploratory analysis.